### ISSUE TITLE
Exception on Windows 7 when running on .Net 4 (No 4.5 installed)

### ISSUE THREAD

**Issue by bjarteskogoy** [cefsharp/CefSharp#1222]
We get a lot of theese on machines with just .Net 4. Upgrading to .Net 4.5 on the same machine seems to have solved the issue.

```
2015-08-26 10:07:15,005 [1] FATAL Activity.App [(null)] - System.ObjectDisposedException: The task has been disposed.
   at System.Threading.Tasks.Task.ThrowIfDisposed()
   at System.Threading.Tasks.Task``1.ContinueWith(Action``1 continuationAction, TaskScheduler scheduler, CancellationToken cancellationToken, TaskContinuationOptions continuationOptions, StackCrawlMark& stackMark)
   at System.Threading.Tasks.Task``1.ContinueWith(Action``1 continuationAction, TaskContinuationOptions continuationOptions)
   at CefSharp.Wpf.ChromiumWebBrowser.OnIsBrowserInitializedChanged(Boolean oldValue, Boolean newValue)
   at CefSharp.Wpf.ChromiumWebBrowser.OnIsBrowserInitializedChanged(DependencyObject d, DependencyPropertyChangedEventArgs e)
   at System.Windows.DependencyObject.OnPropertyChanged(DependencyPropertyChangedEventArgs e)
   at System.Windows.FrameworkElement.OnPropertyChanged(DependencyPropertyChangedEventArgs e)
   at System.Windows.DependencyObject.NotifyPropertyChange(DependencyPropertyChangedEventArgs args)
   at System.Windows.DependencyObject.UpdateEffectiveValue(EntryIndex entryIndex, DependencyProperty dp, PropertyMetadata metadata, EffectiveValueEntry oldEntry, EffectiveValueEntry& newEntry, Boolean coerceWithDeferredReference, Boolean coerceWithCurrentValue, OperationType operationType)
   at System.Windows.DependencyObject.SetValueCommon(DependencyProperty dp, Object value, PropertyMetadata metadata, Boolean coerceWithDeferredReference, Boolean coerceWithCurrentValue, OperationType operationType, Boolean isInternal)
   at System.Windows.DependencyObject.SetCurrentValue(DependencyProperty dp, Object value)
   at CefSharp.Wpf.ChromiumWebBrowser.<CefSharp.Internals.IWebBrowserInternal.OnAfterBrowserCreated>b__25()
   at System.Windows.Threading.ExceptionWrapper.InternalRealCall(Delegate callback, Object args, Int32 numArgs)
   at MS.Internal.Threading.ExceptionFilterHelper.TryCatchWhen(Object source, Delegate method, Object args, Int32 numArgs, Delegate catchHandler)
```

**Comment by amaitland** [comment]
Markup please

**Comment by amaitland** [comment]
Code looks pretty straight forward. Any thoughts?

https://github.com/cefsharp/CefSharp/blob/master/CefSharp.Wpf/ChromiumWebBrowser.cs#L561

Also using the code block markup syntax makes things easier to read, you can see I edited your post for an example.

**Comment by bjarteskogoy** [comment]
Really no idea. I would assume maybe a .Net 4 bug related to something we don't have control over maybe. It has not happened on .Net 4.5, this far. Seems like the task is disposed somehow. Task control between C# and C++ maybe?

**Comment by amaitland** [comment]
> We get a lot of theese on machines with just .Net 4.

When you say a lot, does it happen on every machine?

> We’ve made Tasks usable even after they’ve been disposed.  You can now use all of the public members of Task even after its disposal, and they’ll behave just as they did before disposal. 

Reading http://blogs.msdn.com/b/pfxteam/archive/2012/03/25/10287435.aspx and it looks like they made quite a few subtle changes in `.Net 4.5`.

**Comment by bjarteskogoy** [comment]
I’ve tested on two windows 7 virtual machines with .Net 4.0. It happened consequently on the first browser we intialized. That wasn't handled so our application crashed. I'll try to look into it a bit further in a day or two and see if I can narrow it down to something.

**Comment by cefsharp-ms** [comment]
I have encountered a very similar problem while using the WCF based JS binding implementation. There it was a problem with the CefTaskWrapper implementation, it calls delete in the destructor which will dispose the wrapped task at that point. I'm not sure it is the same issue but might worth a look.

**Comment by amaitland** [comment]
> There it was a problem with the CefTaskWrapper implementation, it calls delete in the destructor which will dispose the wrapped task at that point.

https://github.com/cefsharp/CefSharp/blob/master/CefSharp.Core/Internals/CefTaskWrapper.h#L34

Makes sense that this would have a problem in `.Net 4.0` with the change in behavior outlined above.

**Comment by bjarteskogoy** [comment]
Seems like removing the finalizer on CefTaskWrapper fixes the issue. Would removing this finalizer cause memory leaks?

**Comment by amaitland** [comment]
> Would removing this finalizer cause memory leaks?

Shouldn't do, the `gcroot` template should cleanup gracefully.

**Comment by amaitland** [comment]
I do like to be explicit about such things in general, so to keep things consistent, we could likely just switch to assigning `nullptr`. Something like

```c#
~CefTaskWrapper()
{
	_task = nullptr;
	_scheduler = nullptr;
}
```

**Comment by amaitland** [comment]
That really should happen automatically, so in this case I'm fine with removing the `destructor`

**Comment by amaitland** [comment]
Merged #1239 

Closing now, reopen if required.

**Comment by amaitland** [comment]
Thanks to @cefsharp-ms for the helpful pointer!
